### DISCUSSION TITLE
Network constrained movement

### DISCUSSION THREAD

**Discussion by bamacgabhann** [anitagraser/movingpandas#307] [Ideas]
1. There are still a couple of methods to return distances which don't use the unit conversions, e.g. ```get_length()``` and ```distance()```. Do any of these need the option to return measurements in different units?
2. I'm wondering if it would be worth changing how DISTANCE_COL_NAME and equivalents work. They seem to be working just fine now, but I wondered about the idea of making them lists, with each new distance etc column appending to the list. Checking for existing columns could then be ```if name in DISTANCE_COL_NAME```. Probably doesn't need to be changed, but just thought I'd make the suggestion.
3. On a related note, default names for distance, speed, acceleration cols could be updated to use units, which would allow declaring units without worrying about names. e.g. someone could to ```traj.get_distance(units="km")``` and the column name could return as "distance (km)". Easy to read the CRS units to have a default if someone just does ```traj.get_distance()```. 
4. I am still a bit unsure about the handing of datetimes in different formats, but I'm also unsure about setting them all to numpy datetime64, particularly with pandas moving towards using the pyarrow backend rather than numpy. 
5. More broadly, does anything need to be changed in mpd to allow for the pandas pyarrow backend?
6. Comparison of trajectories - one of the things I've been looking at which might be worth adding to mpd itself is comparison between trajectories. If you have multiple trajectories covering the same route, extract segments where they differ, e.g. in precise route, or in the case I want, speed. Could take two trajectories, or a Trajectory Collection and return segments which differ. 
7. In order to compare trajectories, I think it would be needed to allow reindexing of trajectories to relative time, ie set the first point as 00:00:00 and all subsequent points as timedelta from that, rather than the absolute datetime.
8. Map matching of trajectories would help comparison of trajectories, and could also be useful more broadly. There's already the generaliser, but there's also a few packages which allow map matching. One or more could be imported, or a map matcher could be written in to mpd directly. 
9. Something else which might make trajectory comparison more useful is creation of idealised trajectories within mpd. Given a start and end point, use osmnx or similar to create a route between those points, and turn it into a trajectory with a relative time index which can then be compared to other trajectories on the same route. 
10. I said on one of the issues that I'm working on an alternative faster Stop Detector, I haven't got back to it since then but I will soon. One of the things I intend to add to that (but haven't yet) is the ability to detect stops at specified points, so you could do something like ```detector = mpd.TrajectoryStopDetector(my_traj, gdf)``` and then ```detector.get_stop_points()``` or ```detector.get_stop_time_ranges``` would by default return only the points or time ranges at the locations listed in gdf. 

Would welcome any thoughts on these ideas. The last 5 come from the public transport data I'm currently working on, I am writing something that's specific to that because there's some things I want that wouldn't fit in mpd such as service area analysis etc. and bits that are specific to Ireland. But I thought these ideas above might be useful more broadly in mpd. If you don't think they'd fit, I can just write them for my project of course, but if you think they would be useful to have in mpd, there's no sense in me writing them just in my project code, so I'd be happy to contribute them and take suggestions.

**Comment by anitagraser** [comment]
Thank you for all these thoughts.

1.  Thank you for bringing this up. Yes, I think that these methods should also support conversions to different units to be consistent with `add_distance`, `add_speed`, and `add_acceleration` methods. 
2. If I understand it correctly, this suggestion would support multiple distance/speed/acceleration columns. Do you have a specific use case in mind that motivates this requirement?
3. Yes, with the introduction of different output units, this could be very helpful for the users to avoid confusion. 
4. Unfortunately, I'm not familiar with the potential consequences of this choice.
5. That is an excellent question but I haven't investigated this yet, so I don't know. 
6. To calculate the difference between map-matched trajectory geometries, it may be sufficient to use basic SimpleFeatures difference https://shapely.readthedocs.io/en/stable/manual.html#object.difference which should already be available in GeoPandas through Shapely. Of course, this would not cover the speed case you mentioned. 
7. Yes, depending on the method, some would use relative time and others (such as closest point of approach) would need absolute time.
8. I would prefer to import a well maintained map-matching library over implementing map-matching directly in mpd. If you have any suggestions, feel free to open a feature request ticket. 
9. Similar to 8, I think integrating a routing library could be helpful.
10. Is there a specific spatial tolerance to determine if a detected stop coincides with a point from the input gdf? Or is the input gdf a polygon type? 

In general, 6-9 (and maybe 10) raise the larger issue of if/how we want to support network-constrained movement data. It opens up a whole new can of worms. The first step/hightest priority would probably be a proof of concept of a map-matcher/router integration to clarify the implementation details, e.g. assuming an irregularly sampled GPS track, does the map-matcher just snap the locations the the road graph and return snapped points or does it return the route along the road networks as lines? If lines, how should the timestamps be handled? I suggest to break this thread up by topic to make it easier to keep track: 1-3 on output unit handling; 4-5 on Pandas pyarrow implications; 6-9 on network-constrained movement; and 10 for faster stop detection.

**Comment by bamacgabhann** [comment]
Have split off 1-3 as an Issue #308, 4-5 as a separate discussion #309, and added discussion of 10 to the existing #296; so renaming this for discussion on network-constrained movement.

As a first step, I'll look into map-matching and routing libraries to see if any would work OK.
